distribution/lib/Standard/Table/0.2.32-SNAPSHOT/src/Data/Matching.enso

```diff
@@ -0,0 +1,167 @@
+     list which matched it. If the group of objects matched by a single criterion
+     has more than one element, their relative order is the same as in the
+     input.
+   - name_mapper: A function mapping the provided object to its name which will
+     then be matched with the provided matching criteria. It is set to the
+     identity function by default, thus requiring the input to just be a list of
+     objects to match. But it can be overridden to match more complex objects.
+   - matching_strategy: A Matching_Strategy instance specifying how the
+     criterion should be interpreted.
+   - on_missing: Specifies the behavior when a criterion did not match any
+     objects. By default, a warning is issued, but the operation proceeds.
+     If it is set to `Report_Error`, the operation fails with a dataflow error.
+     If it is set to `Ignore`, the operation proceeds without errors or
+     warnings.
+   - warning_callback: A callback specifying the function to call when issuing a
+     warning. This is a temporary workaround to allow for testing the warning
+     mechanism. Once the proper warning system is implemented, this argument
+     will become obsolete and will be removed. No user code should use this
+     argument, as it will be removed in the future.
+
+   > Example
+     Selects objects matching one of the provided patterns, preserving the input order.
+
+         Matching.match_criteria ["foo", "foobar", "quux", "baz", "Foo"] [".*ba.*", "f.*"] matching_strategy=(Regex case_sensitivity=True) == ["foo", "foobar", "baz"]
+
+   > Example
+     Selects pairs matching their first element with the provided criteria and
+     ordering the result according to the order of criteria that matched them.
+
+         Matching.match_criteria [Pair "foo" 42, Pair "bar" 33, Pair "baz" 10, Pair "foo" 0, Pair 10 10] ["bar", "foo"] reorder=True name_mapper=_.name == [Pair "bar" 33, Pair "foo" 42, Pair "foo" 0]
+match_criteria : Vector Any -> Vector Text -> Boolean -> (Any -> Text) -> Matching_Strategy -> Problem_Behavior -> (Text -> Nothing) -> Vector Any ! No_Matches_Found
+match_criteria objects criteria reorder=False name_mapper=(x->x) matching_strategy=(Exact case_sensitivity=True) on_missing=Report_Warning warning_callback=here.report_warning_placeholder = Panic.recover <|
+    [objects, criteria, reorder, name_mapper, matching_strategy, on_missing, warning_callback] . each Panic.rethrow
+
+    # match_matrix . at i . at j specifies whether objects.at i matches criteria.at j
+    match_matrix = objects.map obj->
+        criteria.map criterion->
+            name = name_mapper obj
+            here.match_single_criterion name criterion matching_strategy
+
+    # Checks if the ith object is matched by any criterion.
+    is_object_matched_by_anything : Integer -> Boolean
+    is_object_matched_by_anything i =
+        match_matrix.at i . any x->x
+
+    # Checks if the jth criterion matches any objects.
```

Comment:
what is "jth"?

Reply:
The `does_criterion_match_anything` is lambda taking an index argument of `j` so this is referring to that.

Assuming that reusing the `i` in the lambda expressions doesn't cause an issue we can change to that for clarity

Reply:
Yeah it is ok with `i` too. I've used `j` so that we always have `i` corresponding to an object and `j` to criteria, so it is easier to reason about the code. But given it is such a local change I guess it's not that important.